Block API: Add role for 'internal' attributes not copied on duplication

stacimc opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->
Closes #29693
Alternative to #32604 

## Description
<!-- Please describe what you have changed or added -->

This PR:
* Adds handling for a new `internal` attribute experimentalRole. Attributes registered with this role will not be copied on block duplication.
* Refactors `__internalWidgetID` to use the new experimental role
 
Example fragment of block.json:

```
            "attributes": {
		"myInternalAttribute": {
			"type": "string",
			"default": "my-default-value",
			"__experimentalRole": "internal"
		}
	},
```

For just one example of how this might be used: the Jetpack `Pay with Paypal` block uses a `productId` attribute to identify the actual product record referenced by the block. This data is meant to be internal to the block and should not be exposed to the user. When the user duplicates the block, we want to copy the other attributes used to store price info and other details, but create a brand new product record/productId.

This is similar conceptually to #32604 with a few changes:
* The use of `internal` terminology rather than `duplicable` or `unique`, which implies additional constraints
* The use of the `__experimentalRole` and associated utilities (thanks @ntsekouras for the suggestion)

## Notes

* In other regards, an attribute with this role acts like a normal attribute
    * ~~It respects any configured default values~~ Default values are not filled in on block duplication either; support for this could be easily added if a viable use case becomes apparent, though
    * It can be managed as normal via `setAttributes`
    * It can apply to any type of attribute (ie it is not necessarily an id field)
* While you can give this role to an attribute with a "source"/"selector" configuration, it is not meaningful to do so. For example, adding it to the `content` attribute of the [Code block](https://github.com/WordPress/gutenberg/blob/a84c037a0e62a344005054102544c34d7b970a6b/packages/block-library/src/code/block.json#L9) would not make sense, because although the attribute would not be copied on duplication, it will populate with the same value from the html.
* Because this is experimental and also relies on experimental features, I have not added documentation; happy to do so if that is appropriate!
* A note about Reusable blocks:
Internal attributes are not removed when you convert a block into a Reusable block. I don’t know a ton about Reusable blocks, so calling this out so that someone more familiar can correct me; my understanding is that multiple instances of a Reusable block are not independent of each other (ie editing one copy affects them all), so internal attributes should also be shared.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
* Play around with adding `__experimentalRole: 'internal'` to any block attribute. 
* Insert the block into a new post, change its attributes, and duplicate it
* Verify that the internal attributes were not copied, but all other attributes were
* ~~Test with attributes that have a `default` value and ensure that the default is respected in the duplicated block~~
* Test Copy+Paste using both the `Copy` menu item, and by keyboard commands
* Test duplicating a block with internal attributes in the widget editor
* Test adding and editing widgets in both the Widget editor and the Customizer
  * Test moving widgets between widget areas/side panels
  * Test a server-side rendered block in the widget editor and make sure the preview is displayed correctly (Basically verify that #29197 is still fixed with the new implementation)

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

stacimc committed:
Add internal experimentalRole for attributes

stacimc committed:
Add tests

glendaviesnz commented:
This works as expected for me when using the block `Duplicate` option but not when using the `Copy` option and pasting the block somewhere else. Should we make the attribute behave the same in duplicate and copy?

andrewserong commented:
Nice approach @stacimc, I like the addition of the `internal` role, and other than the copy/paste behaviour that Glen mentions, it's testing well for me, too!

talldan commented on packages/blocks/src/api/factory.js:
```diff
@@ -32,6 +32,7 @@ import {
 import {
 	normalizeBlockType,
+	__experimentalStripInternalBlockAttributes,
 	__experimentalSanitizeBlockAttributes,
```
@kevin940726 @noisysocks It'd be good to look into whether this PR makes it possible to remove `__experimentalSanitizeBlockAttributes`, and update the widget editor to use an `internal` attribute instead.

noisysocks replied:
Maybe! `__internalWidgetId` can be set on **any** block type, so we'd have to hook into `registerBlockType` and add `widgetId: { type: 'string', __experimentalRole: 'internal' }` to the schema of every single registered block.

If we can do that then that will make `__experimentalCloneSanitizedBlock` unnecessary as this was only added so that duplicating blocks would not copy any attributes that don't exist in the schema e.g. `__internalWidgetId`. See https://github.com/WordPress/gutenberg/pull/29111.

It might be nice to try and do all of that along with this change so that there is a usage of `__experimentalRole: 'internal'` within core. We're less likely to break APIs that we ourselves use 🙂

gziolo replied:
When discussing it in #29693 I envisioned that the API proposed in this PR would be sufficient to refactor the `__internalWidgetId` use case. It'd be great to confirm that.

noisysocks replied:
I can try implementing what I described above tomorrow unless @stacimc wants to give it a try 🙂

stacimc replied:
Unfortunately I think this may require a little more tinkering -- as is, I'm running into the same issue in #29111 where internal attributes get dropped when the user makes changes in the widget editor 🤔 

In fact at the moment I'm seeing the issue even if I keep the separated `cloneBlock` and `__experimentalCloneSanitizedBlock`, because in order to fix the issue mentioned by @glendaviesnz where internal attributes aren't reset when you `Copy` + paste a block, ~~I need to strip the internal attributes within `createBlock` as well. My somewhat tenuous understanding of the widget editor is that it fetches widgets and transforms them into blocks using `createBlock` -- so the internal attributes get reset every time you refresh the page. This ends up causing block invalidation errors.~~

Edit: I missed @ntsekouras' suggestion below which looks like a better alternative 😄 I'll see what I can do with this and take a shot at removing `__experimentalCloneSanitizedBlock` 👍

ntsekouras commented on packages/blocks/src/api/utils.js:
```diff
@@ -296,3 +296,27 @@ export function __experimentalGetBlockAttributesNamesByRole( name, role ) {
+ * @param {Object} attributes The block's attributes
+ * @return {Object} The attributes, stripped of any internal attributes.
+ */
+export function __experimentalStripInternalBlockAttributes( name, attributes ) {
```
I feel this should be something more generic to remove/strip attributes by `role`.

gziolo replied:
Do we have other use cases at the moment? As long as it's an experimental API we don't have to worry about generalization. We can consolidate later.

ntsekouras replied:
No we don't, so for sure leaving it as is, is fine for now 😄 . I just think that this function will change to handle the case for `copy`. We'll see from the new iterations..

ntsekouras reviewed (commented):
Thanks for working on this @stacimc!

IMO this PR could be the place to remove the `__experimentalCloneSanitizedBlock` as mentioned in comments.

What we need to handle this properly is something like the `cloneBlock` without the new `clientId` (that means recursively applying this to innerBlocks). IMO a good  place to add this functionality for `copy` action is here: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/copy-handler/index.js#L124, before the blocks' serialization with something like this:
```
let blocks = getBlocksByClientId( selectedBlockClientIds );
if ( event.type === 'copy' ) {
	blocks = blocks.map( ( block ) =>
		__experimentalRemoveAttributesByRole( block, 'internal' )
	);
}
```

apeatling reviewed (commented):
This tested well for me. The `internal` terminology here feels better. 👍

stacimc committed:
Refactor removeAttributesByRole to take role param and return block

* For extensibility, it can now be used to remove attributes with any
given role
* It takes and returns a block instance; this is so that it will be
easier to use in copy handlers/etc

stacimc committed:
Remove internal attrs when copying via keyboard command

stacimc committed:
Remove internal attrs when copying via CopyMenuItem

stacimc committed:
Do not alter attributes of block being duplicated

`__experimentalRemoveAttributesByRole` takes a block instance and removes
any attributes with the given `role`. When cloning a block, we want to
remove internal attributes on the clone, but not the original copy. This
changes the order of operations in the cloning function to make sure that
those attributes only get removed on the latter.

Without this change, when a block is duplicated in the widget editor all
internal attributes will be stripped from BOTH the copy and the duplicate.

stacimc committed:
Use `__experimentalCloneSanitizedBlock` in copy handlers

When a block is duplicated, we remove internal attributes and then fill in
any default values (when we call `__experimentalSanitizeBlockAttributes`).
When we copy a block, we were only doing the first step of removing the
internal attrs, but not filling the defaults.

This causes a block validation issue, because the serialized block in the
clipboard is missing internal attrs which will get filled in by `createBlock`
when we actually paste.

This commit updates the copy handlers to use the cloneBlock code so that the
behavior matches block duplication exactly. It also reverts the
`__experimentalRemoveAttributesByRole` to the original API that operates on
an attributes object rather than a block instance, to match the signature of
`__experimentalSanitizeBlockAttributes` and clean up the logic in the
cloning function.

stacimc commented:
Some updates:
* I’ve gotten it working for Copy+Paste, both when using keyboard commands and when using the `Copy` menu item
* I’ve fixed a couple bugs that I noticed (more info in commit descriptions):
    * Internal attributes getting stripped from both the copy and the original block in the widget editor
    * Block validation errors caused by not filling in default values for stripped internal attributes on `Copy`
* I’ve taken a look at the `__internalWidgetId` and what we can do to as far as removing `__experimentalCloneSanitizedBlock`. 

Unfortunately I don’t think it’s going to be so easy to get rid of `__experimentalCloneSanitizedBlock`. With internal attributes, we run into the same problem as described in #29111/#28379. That is:
* We don’t want to copy `widgetId` when we duplicate a block; we want to create a new block referencing a separate widget
* On the other hand, we _do_ want to copy `widgetId` in `useBlockSync`, which is using `cloneBlock` to sync external changes (Ref #22639, this is necessary to solve a tricky bug and can’t be changed)

Making `widgetId` into an internal attribute doesn’t actually fix this, at least that I could come up with. We still need `cloneBlock` to strip those attributes in some cases (on duplication) and not in others (useBlockSync). Moreover I don’t think we can even get rid of `__experimentalSanitizeBlockAttributes`, as this is really useful for filling in any default values for internal attributes after they’ve been removed.

That being said, I think it might still be worth making `widgetId` into an internal attribute, because:
* To @noisysocks point, it would be nice to have a usage of the `internal` role in core
* I’d prefer not to accomplish the same thing in two different ways (unregistered attrs, internal attrs)

In my latest commit, I've hooked into `registerBlockType` to register it as an internal attribute with really minimal changes and it seems to be working so far. I'll definitely be updating/adding new tests, but it seems like a good time to check in to see how others feel about the approach. @noisysocks @gziolo does that square with your knowledge of the widgets issues?

stacimc committed:
Use internal attribute for widgetId

stacimc committed:
Update tests for __experimentalRemoveAttributesByRole

noisysocks commented:
> @noisysocks @gziolo does that square with your knowledge of the widgets issues?

Yep, that all makes sense 🙂

noisysocks commented on packages/widgets/src/utils.js:
```diff
@@ -23,11 +27,33 @@ export function getWidgetIdFromBlock( block ) {
+	'blocks.registerBlockType',
+	'core/widget/addAttributes',
+	addAttributes
+);
```
Probably it doesn't affect anything in practice but it might be better to have this `addFilter()` call inside a function which `initialize()` in `edit-widgets` and `customize-widgets` calls. That way merely importing `@wordpress/widgets` doesn't cause something to happen.

noisysocks commented on packages/widgets/src/utils.js:
```diff
@@ -23,11 +27,33 @@ export function getWidgetIdFromBlock( block ) {
+ *
+ * @return {Object} Updated block settings.
+ */
+function addAttributes( settings ) {
```
Can we give this a more specific name? Maybe `addInternalWidgetIdAttribute`?

noisysocks commented on packages/widgets/src/utils.js:
```diff
@@ -23,11 +27,33 @@ export function getWidgetIdFromBlock( block ) {
  * @return {Block} The updated block.
  */
 export function addWidgetIdToBlock( block, widgetId ) {
+	block.attributes.__internalWidgetId = widgetId;
```
What is this change for? Redux and `@wordpress/data` often rely on object reference equality (`foo === bar`) to detect if something has changed, so not sure that it's a good idea to **not** create a new `attributes` object here.

stacimc replied:
No good reason, crossed brain wires as I was thinking about updating an attribute now guaranteed to exist -- you're of course right. Fixed, thanks!

ntsekouras commented:
>Block validation errors caused by not filling in default values for stripped internal attributes on Copy

I haven't checked the changes yet, but in which case would we need to have a default value of an internal property? Wouldn't the absence of this prop be enough for a block's edit function to handle the generation of that prop?

ntsekouras commented on packages/widgets/package.json:
```diff
@@ -29,6 +29,7 @@
 		"@wordpress/core-data": "file:../core-data",
 		"@wordpress/data": "file:../data",
 		"@wordpress/element": "file:../element",
+		"@wordpress/hooks": "file:.../hooks",
```
```suggestion
		"@wordpress/hooks": "file:../hooks",
```

ntsekouras committed:
Update packages/widgets/package.json

stacimc committed:
Move attribute registration to function called in widget initialization

stacimc committed:
Corrected docstring

stacimc commented:
> I haven't checked the changes yet, but in which case would we need to have a default value of an internal property? Wouldn't the absence of this prop be enough for a block's edit function to handle the generation of that prop?

@ntsekouras In the case of `widgetId`, as well as my `productId` use case described in the PR description, that's correct. I was thinking of the `internal` role as something that could be applied to any arbitrary attribute, which would not be copied on duplication but would otherwise function as a normal attribute.

Top of my head I can't think of any great use cases where someone might want an `internal` attribute with a default value, so I'm certainly not attached to keeping it, though. If others can't think of a reason for allowing defaults, I'd be happy to rip out `__experimentalSanitizeBlockAttributes` and update the docs!

stacimc committed:
Check for existence of widget before moving sidebars

When you move a block in the customizer to a different widget sidebar,
we check to see if the block has a widgetId and if so, assume it's a
valid widget that we can just remove from the original sidebar and add
to the destination sidebar. If it's an inner block of a widget, we
instead remove the block from the parent, create a widget from the block,
and add this to the new sidebar. To detect this we expect it to not have a
wwidgetId.

This can break when you take an existing widget and turn it into inner
blocks by transforming it into a Group. The inner block retains its
widgetId through this transformation, but the corresponding widget
instance is removed.

To cover this case, we now check both that the widgetId exists AND that
there is an existing widget with that id.

stacimc committed:
Do not separately sanitize attributes in cloneSanitizedBlock

The only reason we still had this was to fill in default values for
any internal attrs that got stripped on duplication; since there is
no obvious use case for an internal attribute with defaults I'm
removing this for now.

I leave __experimentalSanitizeBlockAttributes as it is still used in
createBlock, and keeping it pulled out like this allows for ease of
unit testing.

stacimc commented:
Added a couple of changes:

**Removed use of __experimentalSanitizeBlockAttributes in __experimentalCloneSanitizedBlock**
I only kept this previously to allow filling in default values for internal attributes that get stripped on cloning, but since I can't think of a reasonable use case for an internal attribute with a default, I'm pulling it out. This means we could technically remove `__experimentalSanitizeBlockAttributes` altogether, because it's only used in one place now (`createBlock`); however I decided to leave it because it makes unit testing neater in my opinion.

**Fixed a bug when moving inner blocks between sidebars in the Widgets section of the Customizer**
Tricky one, documented in [this e2e test](https://github.com/WordPress/gutenberg/blob/b1fd6b2ce6221f1297da78b6c48aeee56130501b/packages/e2e-tests/specs/widgets/customizing-widgets.test.js#L707) -- when you transform blocks into a `Group`, they are ['copied' via `createBlock`](https://github.com/WordPress/gutenberg/blob/b1fd6b2ce6221f1297da78b6c48aeee56130501b/packages/block-library/src/group/index.js#L118) into the inner blocks of the new Group. With the previous `widgetId` implementation, running through `createBlock` like this would drop the `widgetId`, but that doesn't happen with the new implementation.

This caused problems because we were using the absence of the `widgetId` to detect when a block in the widget editor is an inner block and handle it appropriately when moving it between widget sidebars. I've updated it now to check not only that the block has a `widgetId` but also that there is a corresponding widget.

I'm calling this out because it does mean that we can have dangling `__internalWidgetId` attributes on blocks that _used to_ be widgets but were converted to inner blocks. Another solution would be to strip internal attributes on `createBlock` as well as on duplication -- but while this is great for `widgetId`, it's not what we want for internal attributes in general. (Eg: we have some sort of internal `productId`; we don't want that to get stripped from our block if we Group it!)

gziolo commented:
> This means we could technically remove __experimentalSanitizeBlockAttributes altogether, because it's only used in one place now (createBlock); however I decided to leave it because it makes unit testing neater in my opinion.

It doesn't sound like a compelling reason to keep an experimental public API. If possible let's test the same functionality through the public method which is what we care about the most.

stacimc committed:
Remove __experimentalSanitizeBlockAttributes

stacimc commented:
> It doesn't sound like a compelling reason to keep an experimental public API.

Fair point! I've moved the logic back into `createBlock` and cleaned up the tests.

Looks like there's a use in [`server-side-render.js`](https://github.com/WordPress/gutenberg/blob/5430baf854a28cc5a04647129ccd0814997442f1/packages/server-side-render/src/server-side-render.js#L93) that I missed, though. Taking a look 👀

stacimc committed:
Remove unnecessary attribute sanitization in ServerSideRender component

This was previously added because the endpoint called by this component
validates attributes, which would break with the previous implementation of
`__internalWidgetId` (since it wasn't a registered attribute).

With the new implementation, `__internalWidgetId` is a valid attribute and
it's no longer necessary to sanitize here.

glendaviesnz commented:
I gave this another test and seems to be working as advertised for me, eg. I set the Image block height and width attribs to internal and they were not copied over to new blocks using copy or duplicate options:

Before:

![internal-before](https://user-images.githubusercontent.com/3629020/138204884-bc1cc192-0bcd-44d2-8a9d-b49d720fe7b1.gif)

After:

![internal-after](https://user-images.githubusercontent.com/3629020/138204896-b7d2fbeb-66f7-4727-8ed6-2abc1185ddf5.gif)


Also worked in Widget screen and the Archive server side block also seemed to work as expected there.

Tropicalista commented:
Any ideas on when this will be merged?

aaronrobertshaw commented:
Nice work @stacimc. This is tests as advertised for me 👍 

I tested using the following blocks and internal attributes;

- Media & Text -  `mediaId`, `mediaUrl`, `mediaLink`, and `mediaType` as internal attributes
- Gallery - `columns` as internal an attribute

✅ Inserting block and updating still functions as per normal
✅ Duplicating the block with internal attributes works as advertised, only the internal attributes were cleared
✅ Copy and pasting a block, whether by the menu item or keyboard shortcuts, is consistent with duplicating
✅ Duplicating blocks within the Widget editor and customizer works for me
✅ Dynamic blocks in the Widget editor continue to preview ok (tested with Page List and Archive)

I'm probably lacking too much context to critique the approach but it makes sense to me. I'll defer to others regarding final approval.

gziolo commented on packages/block-editor/src/components/block-settings-menu/block-settings-dropdown.js:
```diff
@@ -33,7 +37,12 @@ const POPOVER_PROPS = {
+		blocks = blocks.map( ( block ) =>
+			__experimentalCloneSanitizedBlock( block )
+		);
+		return serialize( blocks );
```
Have you explored whether we could move `__experimentalCloneSanitizedBlock` to the `serialize` method? I see a similar logic together in another file in this PR. The way it works now is fine but we will have to ensure that we use the same logic any time we add new functionality or what’s worse when plugins use the current public API. I don’t know if that’s possible but it would be a nice improvement for developers.

stacimc replied:
I see your point, but I don't think we can move the logic up any higher. If we move `__experimentalCloneSanitizeBlock` into `serialize`, we would end up stripping out internal attributes on much more than duplication, every time the post is reloaded.

In this file and in the copy-handler, I think it makes sense that when you want to make a copy of a block you should explicitly clone it before serializing. There is some danger that this could be missed with new functionality that involves block duplication, but I can't think of a way of enforcing it.

gziolo commented:
The feedback received so far looks very encouraging 💯

@ntsekouras, can you have another look at the final proposal and whether it aligns what you had in mind?

@kevin940726 or @noisysocks, can you confirm that it integrates and works correctly with the widgets?

gziolo commented on packages/widgets/src/utils.js:
```diff
@@ -31,3 +35,32 @@ export function addWidgetIdToBlock( block, widgetId ) {
+	return {
+		...settings,
+		attributes: {
+			...settings.attributes,
```
What’s the reason it gets injected with a filter rather than included in the block settings? It would be great to document it in this callback so it doesn’t get accidentally inlined in `block.json`.

stacimc replied:
The filter lets us inject the `widgetId` in the appropriate context, rather than to all blocks even outside of the widget editor. This used to be absolutely necessary because we used the existence of a `widgetId` attribute to determine whether a block was a widget, but I actually think that's not strictly necessary after[ these changes](https://github.com/WordPress/gutenberg/pull/34750/commits/39eaea7ffd71a50cd8009e6ffd087639202fa1ed).

noisysocks replied:
> but I actually think that's not strictly necessary after these changes.

That link doesn't work for me anymore. I imagine it's still necessary that we add `widgetId` to all blocks using this filter because `widgetId` is how the widgets editor knows which widget entity to update when a block is modified.

stacimc replied:
Sorry, link broken after a force push :) [Updated link,](https://github.com/WordPress/gutenberg/pull/34750/commits/0c4673b7a1e74d9414aec63146cd8c9c10afb068) although I think that point is moot -- I meant that it used to be necessary that `widgetId` only be applied to blocks that actually map to a widget entity rather than to all blocks, but really it only matters if the attribute has a value set.

Regardless, I think it's better to add `widgetId` using this filter than automatically to all blocks regardless of context, unless I misunderstand.

noisysocks commented on packages/block-editor/src/components/block-settings-menu/block-settings-dropdown.js:
```diff
@@ -33,7 +37,12 @@ const POPOVER_PROPS = {
-	const ref = useCopyToClipboard( () => serialize( blocks ), onCopy );
+	const ref = useCopyToClipboard( () => {
+		blocks = blocks.map( ( block ) =>
+			__experimentalCloneSanitizedBlock( block )
```
It's my first time looking at this code in a few months and, reading this, I struggled to grok what `__experimentalCloneSanitizedBlock` does. In particular I didn't really understand it's relationship to `cloneBlock`.

I think that a better API might be to add optional settings to `cloneBlock`.

```js
cloneBlock( block, null, null, { includeInternalAttributes: false } )
```

This makes the parallels to `cloneBlock` very explicit. It's also clearer, I think, to avoid the word "sanitize" which could mean all sorts of things.

These settings could also exist on `serialize` which would address @gziolo's comment.

```js
serialize( blocks, { includeInternalAttributes: false } )
```

(Not a blocking comment—we can absolutely iterate on the API later.)

stacimc replied:
I really like this suggestion, especially because it provides a better answer than I was able to give @gziolo re: serialization. And +1 for concerns on 'sanitize'; I struggled to come up with a better name and would be much happier to avoid it!

noisysocks commented on packages/blocks/src/api/utils.js:
```diff
@@ -300,3 +257,29 @@ export function __experimentalGetBlockAttributesNamesByRole( name, role ) {
+ *
+ * @return {Object} The attributes, with those matching the `role` removed.
+ */
+export function __experimentalRemoveAttributesByRole( name, attributes, role ) {
```
This function feels a bit too specialised to me. It's better that complex functionality is achieved by composing together smaller functions rather than adding a new function. That way we're not shipping unnecessary bytes to the browser or having to maintain a large public API. 

Could callers use `_.omit` instead?

```js
const filteredAttributes = omit(
	block.attributes,
	getBlockAttributesNamesByRole( block.name, 'internal' )
);
```

https://devdocs.io/lodash~4/index#omit

tyxla replied:
Side note: I know this is a rather old comment, but just in case it helps, I wanted to point out that since #52571 we're no longer using Lodash in the repository, and that means we can use native language functionality for omission. See #43711 and #43704 for examples of static and dynamic omission of properties from an object.

noisysocks commented on packages/customize-widgets/src/filters/move-to-sidebar.js:
```diff
@@ -55,10 +55,10 @@ const withMoveToSidebarToolbarItem = createHigherOrderComponent(
-
-			if ( widgetId ) {
+			const activeSidebarAdapter = activeSidebarControl.sidebarAdapter;
+			if ( widgetId && activeSidebarAdapter.getWidget( widgetId ) ) {
```
What's this change do?

stacimc replied:
This was necessary to [fix a regression](https://github.com/WordPress/gutenberg/pull/34750/commits/0c4673b7a1e74d9414aec63146cd8c9c10afb068).

When you move a block in the customizer to a different widget sidebar, previously we would check to see if the block has a widgetId and if so, assume it's a valid widget that we can just remove from the original sidebar and add to the destination sidebar. We would detect things like inner blocks by the absence of a widgetId (necessary in order to handle inner blocks differently.)

With the new implementation this can break when you take an existing widget and turn it into inner blocks by transforming it into a Group. The inner block retains its widgetId through this transformation, but the corresponding widget _instance_ is removed.

To cover this case, we now check both that the widgetId exists AND that there is an existing widget with that id.

noisysocks commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -88,20 +87,12 @@ export default function ServerSideRender( props ) {
 
-		const sanitizedAttributes =
-			attributes &&
-			__experimentalSanitizeBlockAttributes( block, attributes );
```
Why were we previously sanitising the attributes here?

stacimc replied:
Previously we needed this because the endpoint called by this component validates attributes, which would break with the previous implementation of `__internalWidgetId` (since it wasn't a registered attribute).

With the new implementation, `__internalWidgetId` is a perfectly valid attribute and it's no longer necessary to sanitize here.

stacimc replied:
Although after rebasing I'm actually getting those validation errors again when I test an SSR block in the widget editor -- so maybe I spoke too soon! Taking a look to see what's changed, as I would still expect this to work without sanitization.

noisysocks reviewed (commented):
Sorry it's taken so long to get this reviewed!

The code looks really good. I left a few questions that I had while reviewing and a few non-blocking comments about potentially making the public API a bit clearer. Happy to explore that as a follow up if you'd prefer.

I'll give this a final test in my tomorrow morning, though it looks like @aaronrobertshaw has already done a good job in that department.

stacimc committed:
Remove removeAttributesByRole

stacimc commented:
Thanks for taking another look, @noisysocks! This has been rebased and the unnecessary `removeAttributesByRole` refactored.

After rebasing I've re-tested and it seems that server side blocks are broken again in the Widget editor & customizer (I'm testing with the Archive block). I'm certain this was working before the rebase, but now [block validation is failing](https://core.trac.wordpress.org/browser/tags/5.9/src/wp-includes/rest-api/endpoints/class-wp-rest-block-renderer-controller.php#L71) due to missing the `__internalWidgetId` attribute in the block schema on the backend.

I'm not sure why this is no longer working -- I'll keep digging, but in the worst case I think we might have to reintroduce `__experimentalSanitizeBlockAttributes` to keep the fix from #29197.

noisysocks commented:
> After rebasing I've re-tested and it seems that server side blocks are broken again in the Widget editor & customizer (I'm testing with the Archive block). I'm certain this was working before the rebase, but now [block validation is failing](https://core.trac.wordpress.org/browser/tags/5.9/src/wp-includes/rest-api/endpoints/class-wp-rest-block-renderer-controller.php#L71) due to missing the `__internalWidgetId` attribute in the block schema on the backend.

Hm, that's strange. To be honest I'm not sure why it would have ever worked. We're only declaring the `__internalWidgetId` attribute in the client which means the server can't possibly know about it. 

It makes intuitive sense to me that `ServerSideRender` should remove _internal_ attributes before sending them to the server, since our definition of _internal_ is something like _attribute that exists only in client side memory_. 

Keeping the fix that was added in https://github.com/WordPress/gutenberg/pull/29197 makes sense. I think I'd prefer to still remove `__experimentalSanitizeBlockAttributes`, because the word _sanitize_ is very ambiguous to me 😀

stacimc commented:
>  To be honest I'm not sure why it would have ever worked. 

Now I've been looking at it closer, that's exactly what's confusing me! I'm actually wondering if we all tested inserting a dynamic block but not updating or re-rendering it.

> Keeping the fix that was added in #29197 makes sense. I think I'd prefer to still remove __experimentalSanitizeBlockAttributes, because the word sanitize is very ambiguous to me 

+1, I can just omit internal attributes using `__experimentalGetBlockAttributesNamesByRole` here as well so it's much more explicit.

> It makes intuitive sense to me that ServerSideRender should remove internal attributes before sending them to the server, since our definition of internal is something like attribute that exists only in client side memory.

I'm not sure about this. It makes a lot of sense in the case of the `widgetId`, but 'normal' attributes registered in `block.json` could easily be marked as internal as well. One of the driving use cases for this work is a Payments type block that wants to store a `productId` in an attribute, but needs to make sure that it is not copied on block duplication.

noisysocks commented:
> I'm not sure about this. It makes a lot of sense in the case of the `widgetId`, but 'normal' attributes registered in `block.json` could easily be marked as internal as well. One of the driving use cases for this work is a Payments type block that wants to store a `productId` in an attribute, but needs to make sure that it is not copied on block duplication.

Yeah that's a good point. Some loose thoughts...

1. I'm not sure if there's an actual reason [we disallow additional properties](https://github.com/WordPress/wordpress-develop/blob/3707a6ca5ceb3dc0f71d173ef7660dd9c8bd3f72/src/wp-includes/rest-api/endpoints/class-wp-rest-block-renderer-controller.php#L68) in the render endpoint. I don't see a great harm in having the REST API not throw an error if extra attributes are given which don't exist in the block schema.

2. We could add more  _roles_. Maybe _internal_ means that it will be shared with the server but _private_ means that it will remain on the client... or something. It could become (already is?) very confusing.

3. Perhaps _role_ is the wrong abstraction for this. Maybe we should be explicit about what each attribute does and doesn't support.

   ```js
   "productId": {
   	"type": "string",
   	"supports": {
   		"serialize": true, // whether serialize() will strip out the attribute
   		"copy": false, // whether copying/duplicating will strip out the attribute
   		"serverSideRender": false, // whether ServerSideRender will strip out the attribute
   	},
   }
   ```

4. Maybe a declarative API is the wrong choice, as we can't anticipate all use cases. Perhaps blocks ought to be able to define a function that implements exactly what needs to happen when the block is copied or duplicated.

   ```js
   edit: BlockEdit,
   save: BlockSave,
   copy( { name, attributes, innerBlocks } ) {
   	return createBlock( name, omit( attributes, [ 'productId' ] ), innerBlocks );
   },
   ```

gziolo commented:
@noisysocks, some good thinking. @dmsnell started a discussion on a similar topic in https://github.com/WordPress/gutenberg/discussions/38191. It isn't exactly that, but it also tries to address the limitation of the current shape of the attributes.

We also discussed in the past other possible roles in https://github.com/WordPress/gutenberg/issues/29693#issuecomment-915079315. Maybe we need presets that are collections of characteristics that describe what can happen to an attribute when a given action happens? Individual behaviors would have to be defined, but some of them can be digested from the use case we deal with like:
- can be copied
- has to be unique
- can be exposed to the frontend/client

noisysocks commented:
Thanks for connecting those dots @gziolo! What @youknowriad says in https://github.com/WordPress/gutenberg/discussions/38191#discussioncomment-2041269 is very relevant and sways me away from a `copy` function.

> The server might also need to know about this information and declaring this as a JS function forbids that entirely. So I'm actually more in favor of the opposite direction (assign a "role" to an attribute)

---

I think I'm leaning towards renaming `"role": "internal"` to `"role": "identifier"` (or `"foreignKey"`?) as that describes, semantically, what the attribute **is**.

We would then have two roles:

- `"content"` — indicates that the attribute the block's primary text content.
- `"identifier"` (or `"foreignKey"`?) — indicates that the attribute is a reference to some external resource.

Then, to fix the problem identified in https://github.com/WordPress/gutenberg/pull/34750#issuecomment-1027363517:

1. We could just set `'additionalProperties'` to `true` on the server. I don't really see the need to be so strict there. This might require a Core patch, though, as I'm not sure that it's possible for a plugin to re-register a route.

2. Or, we could have `ServerSideRender` perform a `GET` on the `/block-types` endpoint before making its request to `/block-renderer`. If an attribute isn't in the `attributes` object that comes back from `/block-types`, then it hasn't been defined on the server, and so we must not send it to `/block-renderer`.

Sorry to derail the conversation @stacimc—you've put your foot into some interesting problems here 😀

stacimc commented:
> Sorry to derail the conversation @stacimc—you've put your foot into some interesting problems here 😀

@noisysocks Not derailed at all, I think it's important to get this right!

> I think I'm leaning towards renaming "role": "internal" to "role": "identifier" (or "foreignKey"?) as that describes, semantically, what the attribute is.

My concern with this is that ultimately the goal of this PR is to provide a way to make an attribute that isn't copied on block duplication -- and I'm not sure it's obvious or desirable that an attribute with the `"identifier"` role would always behave that way. You might also have other attributes which _aren't_ foreign keys, but which you'd still like to prevent being copied.

I don't think it's perfect but I'd still lean towards `internal` as a better descriptor of the behavior we'd expect from an attribute with this role. I wouldn't assume that an `internal` attribute has expectations about privacy/shouldn't be shared with the server, unlike some more loaded terms like `private` or `unique`.

> We could just set 'additionalProperties' to true on the server. I don't really see the need to be so strict there. This might require a Core patch, though, as I'm not sure that it's possible for a plugin to re-register a route.
>
> Or, we could have ServerSideRender perform a GET on the /block-types endpoint before making its request to /block-renderer. If an attribute isn't in the attributes object that comes back from /block-types, then it hasn't been defined on the server, and so we must not send it to /block-renderer.

Would you prefer this approach to bringing back `__experimentalSanitizeBlockAttributes`? This PR didn't introduce that API or include it in `ServerSideRender`, so it isn't a new addition. We could always remove it later if we're able to set `additionalProperties` on the server.

andrewserong commented:
> I don't think it's perfect but I'd still lean towards internal as a better descriptor of the behavior we'd expect from an attribute with this role.

Apologies if I'm missing context, but I quite like the idea of `internal`, too. While a semantic description could be useful, I think in this case, it sounds like it might be more important to flag how the attribute is used, rather than what it is? One of the ideas I had in the back of my mind is that we might want to one day use a `public` role for attributes, where we might expose those attributes or render them inline in `data-my-attribute` attributes in rendered HTML so that a consumer of post content via the REST API could then grab those attributes via JS and then do something with them (e.g. possibly to regenerate / rebuild block styles or add other dynamic block behaviour). And there's a nice symmetry with `public` and `internal`.

That's clearly well outside the scope of this PR, but just thought I'd share the potential future use case in case it helps with deciding on naming!

noisysocks commented:
> My concern with this is that ultimately the goal of this PR is to provide a way to make an attribute that isn't copied on block duplication -- and I'm not sure it's obvious or desirable that an attribute with the `"identifier"` role would always behave that way. You might also have other attributes which _aren't_ foreign keys, but which you'd still like to prevent being copied.
>
> I don't think it's perfect but I'd still lean towards `internal` as a better descriptor of the behavior we'd expect from an attribute with this role. I wouldn't assume that an `internal` attribute has expectations about privacy/shouldn't be shared with the server, unlike some more loaded terms like `private` or `unique`.

> While a semantic description could be useful, I think in this case, it sounds like it might be more important to flag how the attribute is used, rather than what it is?

It seems like maybe we have two concepts: _role_ (_content_, _identifier_, _styling_, etc.) which gives semantic meaning to an attribute, and _access_ (_public_, _internal_, _private_) which controls where an attribute can be used (client, server, serialised HTML, etc.)

I don't really think _public_, _internal_, _private_, etc. are intuitive, though. Do you think it would be clearer to specify precisely what behaviours are desired? For example, `"copy": false` on an attribute could mean that it isn't copied.

> Would you prefer this approach to bringing back `__experimentalSanitizeBlockAttributes`? This PR didn't introduce that API or include it in `ServerSideRender`, so it isn't a new addition. We could always remove it later if we're able to set `additionalProperties` on the server.

Yeah we can bring back `__experimentalSanitizeBlockAttributes`. Better that we iterate on this API in subsequent changes instead of trying to get it perfect the first try in this PR.

dmsnell commented:
Even with the brief discussion here it's clear that singular roles are an over-simplification of an attributes relevance in various contexts. At least having a set of roles that an attribute participates in could go a long way to helping us build a reliable interface for attribute selection.

Granted, it's clear that these roles can be dependent upon the values of the attributes, but at least as far as describing an attribute's role goes specifying a set of roles or context in which that attribute carries weight goes pretty far.

Maybe it could resolve some of the inherent tension behind trying to figure out what exactly and attribute _is_ by asking instead "do you do this too?" These are contexts in which the attributes are relevant or possibly capabilities each attribute has.

```js
"attributes": {
	"myInternalAttribute": {
		"type": "string",
		"default": "my-default-value",
		"__experimentalRoles": {} // an empty set means we participate in no roles
	},
	"myNormalAttribute": {
		"type": "string", // no property set means we participate in everything
	},
	"produceId": {
		"type": "number",
		"__experimentalRoles": { "copy": false } // everything except false values
	},
	"styleThing": {
		"type": "string",
		"__experimentalRoles": { "style": true, "copy": true } // only these things
	}
},
```

stacimc commented:
> It seems like maybe we have two concepts: _role_ (_content_, _identifier_, _styling_, etc.) which gives semantic meaning to an attribute, and _access_ (_public_, _internal_, _private_) which controls where an attribute can be used (client, server, serialised HTML, etc.)
> 
> I don't really think _public_, _internal_, _private_, etc. are intuitive, though. Do you think it would be clearer to specify precisely what behaviours are desired? For example, `"copy": false` on an attribute could mean that it isn't copied.

My initial idea for this work was along those lines -- I actually opened an earlier PR (#32604) that added a `duplicable` property, so you could configure: `"duplicable": false`. At the time we decided to go with using `role` instead because it made use of existing utilities and didn't require adding an extra attribute property.

I like `copy` a lot as an alternative for clarity -- especially if we intend to add `public` in the future, since `public/internal` feel like they would be related, when we're not really concerned with access 🤔 

> Even with the brief discussion here it's clear that singular roles are an over-simplification of an attributes relevance in various contexts. At least having a set of roles that an attribute participates in could go a long way to helping us build a reliable interface for attribute selection.

I think this would help considerably. It could possibly get a bit tricky if an attribute attempts to configure multiple roles that cannot be simultaneously held, but it's definitely clear that limiting an attribute to a single role is insufficient if we're going to use `role` to describe a mix of semantic and behavioral properties.

noisysocks commented:
> Even with the brief discussion here it's clear that singular roles are an over-simplification of an attributes relevance in various contexts.

Agreed. It's difficult to draw neat boxes around different behaviours, and, even when you do, naming them in a way that makes it clear to extenders what those behaviours are is very difficult.

> Maybe it could resolve some of the inherent tension behind trying to figure out what exactly and attribute _is_ by asking instead "do you do this too?"

Agreed. This feels more flexible and future proof to me. It's analogous to what we already do with block definitions with `supports`.

> My initial idea for this work was along those lines -- I actually opened an earlier PR (#32604) that added a `duplicable` property, so you could configure: `"duplicable": false`. At the time we decided to go with using `role` instead because it made use of existing utilities and didn't require adding an extra attribute property.

I think `role` may have turned out to be a red herring 😬 What do you think about returning to your earlier approach? We can apply some of what we learned here by renaming `unique` to something like `supports.copy` and perhaps try to get some code-reuse by turning `getBlockAttributesNamesByRole` into something more generic, e.g.

```json
"attributes": {
	"content": {
		"type": "string",
		"role": "content",
	},
	"productId": {
		"type": "string",
		"supports": {
			"copy": false,
		},
	},
}
```

```js
const contentAttributes = filterBlockAttributes( block.name, block.attributes, { role: 'content' } );
const copyableAttributes = filterBlockAttributes( block.name, block.attributes, { supports: { copy: true } } );
```

stacimc commented:
Since it's a different approach to the API, I've opened up #38643 as an alternative to this PR following suggestions by @noisysocks 😄 

That PR removes the use of `__experimentalRole` and instead adds a `supports` object:

```
"attributes": {
	"content": {
		"type": "string",
		"role": "content",
	},
	"productId": {
		"type": "string",
		"supports": {
			"copy": false,
		},
	},
}
```

ntsekouras commented:
Sorry for catching up late with this one and thanks for all the work here @stacimc!

>It seems like maybe we have two concepts: role (content, identifier, styling, etc.) which gives semantic meaning to an attribute, and access (public, internal, private) which controls where an attribute can be used (client, server, serialised HTML, etc.)

For me the above feels what's the case here. I don't think we should try to mix two different things in one API. For me the use case we need here is to have a way to mark an attribute as unique and handle everything that this entails, like `copy, duplication`, etc.. That's why I had suggested to use something like `role: unique|identifier` or something similar. 

Why do we need to create a new `supports` API with `copy`? What's the difference from a new role and does it bring more value? Do we expect many other supports? Could this lead to an overly declarative (even bloated) blocks API?

I think that many discussions here have been triggered by the `widgetId` case which if I understand correctly didn't declare the attribute. Is this something though that could dictate the augmentation of blocks API with semantics?

I'll catch up with the code now, but that were my first thoughts after reading all the discussions.

dmsnell commented:
> Could this lead to an overly declarative (even bloated) blocks API?

This is definitely a concern I had which led me to propose a function-based approach. For one, I imagine many blocks want to look at the values to determine these things; two, it seems like something where the potential space of roles and context these things play can kind of explode in complexity if defining that with a DSL.

The idea of having both seems workable to me: either you have a list of relevant roles for an attribute _or_ you supply a function to work out the attribute sets.

Even with a DSL here we're not going to have the problem solved on the server since attributes still need to be sourced from the block's HTML. It would only be able to be relied upon in the server if coupled to a `block.json` file with the attributes defined, and even then we're going to have contexts (such as undo) which don't make much sense on the server.

stacimc commented:
Thanks @ntsekouras, @dmsnell.

> I think that many discussions here have been triggered by the widgetId case which if I understand correctly didn't declare the attribute.

The `widgetId` has a similar problem to my original `productId` use case -- it's an attribute that shouldn't be copied on block duplication -- and was brought in so we would have an established use case of the new API.

> That's why I had suggested to use something like role: unique|identifier or something similar.
>
> Why do we need to create a new supports API with copy? What's the difference from a new role and does it bring more value? Do we expect many other supports? Could this lead to an overly declarative (even bloated) blocks API?

The problem I've been having is finding a semantic 'role' that strictly describes the behavior (attribute isn't copied on block dupe/copy).
   * `unique` is a loaded term that suggests actual uniqueness across instances will be enforced
   * `identifier` is a good semantic descriptor for the `productId` and `widgetId` examples but might not imply the behavior we're implementing (should we assume all identifiers shouldn't be copied when you duplicate a block? What if my attribute isn't an identifier, but I still don't want it duplicated?)
   * `internal` works _pretty_ well but might imply something about access, especially alongside roles like `public`/`private`
   * `copy` describes the behavior very explicitly, but at this point are we overloading the meaning of a `role`?

We can support multiple `__experimentalRoles` for a single attribute, but if each role is permitted to define its own set of behaviors (copy/undo/etc) for an attribute, how do we handle the interaction of conflicting roles? Discrete `supports` might be easier to reason about in that case.

dmsnell commented:
> Discrete supports might be easier to reason about in that case.

for what it's worth I find `supports` a tighter fit to most of the discussion here. in my own thinking I consider "relevance" to "contexts" as the factor. This is why I have trouble understanding fixed static roles for each attribute.

I wouldn't expect `roles` to have its own `supports`, but either could work.

> `identifier`

This reads to me more like attributes that identify the block, which I think would normally imply that they copy when duplicated instead of the inverse. Could these be other things that aren't ids that serve the same purpose which could help us figure out a good name for them?

 - a page in a book
 - serial number
 - part number
 - day of year

Maybe these are some kind of `reference` or `xref` or `cross-ref`? They do uniquely identify a block but they are some kind of selection inside other block settings?

The concept of `identity` here is vague.

ntsekouras commented:
>The idea of having both seems workable to me: either you have a list of relevant roles for an attribute or you supply a function to work out the attribute sets.

This is not flexible as functions aren't serialisable and we need to have access to things like that server side and sometimes we need to register things server side. The same problem occurred for `isActive` function for block variations where we needed to define dynamically(server side) variations. We ended up having `js filters` for that function([example here](https://github.com/WordPress/gutenberg/pull/29095/)).

>The problem I've been having is finding a semantic 'role' that strictly describes the behavior (attribute isn't copied on block dupe/copy).
>The concept of identity here is vague.

I can see your points and maybe it's okay to try with `supports` but make it experimental or unstable for start and only for `copy`.

For the record the use case that made me look into this problem more is the `Query Loop` `queryId`. It needs to be unique because it's used for the associated pagination to work properly.

So I guess I'm going to check your other PR 😄

Tropicalista commented:
Will this take in account also the removal of unique attributes when we export content?

stacimc commented:
> Will this take in account also the removal of unique attributes when we export content?

I'm leaving this open for the moment as discussion is ongoing, but it's likely this PR will be closed in favor of #38643. That PR removes the use of `internal`/`unique` and instead provides a `supports` object, with support for a `copy` key:

```
"attributes": {
	"productId": {
		"type": "string",
		"__experimentalSupports": {
			"copy": false,
		},
	},
}
```

So to answer your question @Tropicalista I think that approach would only handle removing those attributes on block copy/duplication -- but an `export` support could theoretically be added as another key to `__experimentalSupports`. In fact I think that's a point in favor of that approach.

youknowriad commented:
>  but an export support could theoretically be added as another key to __experimentalSupports. In fact I think that's a point in favor of that approach.

I actually think it's a point in favor of the current PR because with the "supports", we'd have to think and add a flag for everyone case where the current PR adds meaning/semantics and not be dependent on specific use-cases. We can make informed decisions for each use-case depending on the role of the attribute.

noisysocks commented:
@youknowriad: What "role" do you think something like `productId` or `widgetId` would have? It's been difficult to define, see https://github.com/WordPress/gutenberg/pull/34750#issuecomment-1035419119.

youknowriad commented:
@noisysocks I see the point, I think it's not clear when you say this blocks is identified by this `productId` whether this block should be unique or not. So I think `identifier` doesn't mean `unique` and `unique` doesn't seem to be a "role" but more its own property of the attribute: `unique: true` that could imply a number of restrictions on the editor:

 - What happens when you copy/export...
 - What happens if the uniqueness is not guaranteed (say I modified the HTML of the post), does it show a message?...

To clarify, I'm not totally against `supports` either, just sharing my thoughts to help you all make the right decisions.

stacimc commented:
> doesn't seem to be a "role" but more its own property of the attribute: `unique: true` that could imply a number of restrictions on the editor

@youknowriad The earliest implementation of this work (https://github.com/WordPress/gutenberg/pull/32604) was actually exactly that!

```
{
    productId: {
        type: 'number',
        unique: true,
    }
}
```

The concern was that `unique` is quite a loaded term, and implies much more than we need (for the `productId` example, it's even possible you might want to manually create multiple blocks linking the same underlying product). Some alternative name suggestions explored were `internal` and `duplicable`. The former is a bit unclear, but maybe `duplicable` still has promise?

youknowriad commented:
> for the productId example, it's even possible you might want to manually create multiple blocks linking the same underlying product

Shouldn't you be able to duplicate and keep the same productId if that's possible?

ntsekouras commented:
>for the productId example, it's even possible you might want to manually create multiple blocks linking the same underlying product

I agree with Riad. I think in the above use case we don't need any semantic marking. It's just the default behavior for block attributes. Am I missing something?

stacimc commented:
> Shouldn't you be able to duplicate and keep the same productId if that's possible?

I was just pointing out why the name `unique` is [misleading](https://github.com/WordPress/gutenberg/pull/32604#issuecomment-860630137)/undesired for some use cases. Ie: I want a way to configure an attribute such that it is not automatically copied when I copy or duplicate the block, but I have no expectation that _uniqueness_ will be enforced otherwise. To continue using my product example: maybe in the future I want to allow users to update the `productId`.

Other suggestions were `internal`, `duplicable`, `copy`.

```
{
    productId: {
        type: 'number',
        duplicable: false,
    }
}
```

noisysocks commented:
Yes, to me, `unique` implies more than what will actually happen which is no good.

I still like `supports.copy` as it describes minimally the behaviour that setting it will cause. This makes it easier to compose different behaviours together. For example, one day we might have a whole suite of customisable attribute behaviours.

```js
{
    attributeName: {
        type: 'number',
	supports: {
	        copy: true, // whether attribute remains when block is copied or duplicated
		serialize: true, // whether attribute remains when block is serialised to HTML
		restApi: true, // whether attribute is shown in the REST API
		undo: true, // whether changes to the attribute are undoable / redoable
	},
    }
}
```

At any rate, I feel the discussion is going in circles a bit and that we ought to lean on `__experimental` instead of our prescience. I'm happy with https://github.com/WordPress/gutenberg/pull/38643 😀
